Python HTTP Server Example

Depetrol opened the pull request.

edwardalee commented on examples/Python/src/WebServer/logging.lf:
```diff
@@ -0,0 +1,136 @@
+  import asyncio
+  import uuid
+  import time
+  import random
```
Is random used?  Is time used? uuid?

Depetrol replied:
You are right, random and time is not used. I've removed them from the preamble. `uuid` is used to generated unique ids for each request.

edwardalee reviewed (approved):
LGTM.  At some point, it would be good develop the WebServer reactors into a reusable library reactor, as done in the C playground, but this is a good starting point for such a project.  I've suggested a bunch of very minor editing changes and raised some questions.

Depetrol commented on examples/Python/src/WebServer/README.md:
````diff
@@ -0,0 +1,76 @@
+
+The handler is as follows:
+
+```python
+@self.app.post("/addone")
+async def addone(request: Request):
+    event = asyncio.Event()
+    request_id = str(uuid.uuid4())
+    self.events[request_id] = event
+    num = int((await request.json())["data"])
+    addone_action.schedule(0, [request_id, num])
+    await event.wait()
+    num = self.events[request_id]
+    del self.events[request_id]
+    return {"status": "success", "num": num}
+```
````
I've updated the documents to make this clearer.

Depetrol commented:
Thank you for your detailed comments! I implemented a reusable library reactor `WebServer` that can handle HTTP requests. I had difficulty abstracting the reusable reactor because of routing issues, and I found that this could be bypassed by sending the routing information as payload instead of using different endpoints. I also reimplemented both examples(+1 and logging) using the reusable library reactor, and I think the code has cleaned up quite a bit. Some documentation was also added.

edwardalee commented on examples/Python/src/WebServer/README.md:
````diff
@@ -0,0 +1,113 @@
+    return {"status": "success", "num": num}
+```
+
+`self` here refers to the `WebServer` reactor in which the handler is defined. The `self.app` is an instance of `FastAPI` application instance, defined as a state of the `WebServer` reactor. Import statements are in the `preamble` and not shown here for simplicity. This handler function will be triggered to generate a response to an HTTP `POST` request at the `/addone` endpoint.
````
Is "endpoint" the right word here?  I would have called in the "path".  I think "endpoint" may refer to the entire URL. If you agree, perhaps the parameter in the WebServer reactor should be changed too.

Depetrol replied:
Yes, I think "path" might be a better word. I'll change the code as well!

edwardalee commented on examples/Python/src/WebServer/logging.svg:
This SVG file seems clipped.

Depetrol replied:
I think this a GitHub svg file display issue, it seems to be rendering correctly in markdown.

edwardalee commented on examples/Python/src/lib/WebServer.lf:
```diff
@@ -0,0 +1,84 @@
+ *
+ * The `endpoint` parameter is the endpoint at which the server listens for requests.
+ *
+ * The `request` output is a tuple of two values: the request ID and the request data.
```
Can you clarify what is the request data?  I assume this is the body of the POST, which is normally JSON (but I guess it can be any string?).

edwardalee reviewed (approved):
The library reactor looks great!  I added a couple more questions, but otherwise this is ready to merge.

lhstrh commented on examples/Python/src/lib/WebServer.lf:
```diff
@@ -0,0 +1,84 @@
+ *
+ * The `response` input is a tuple of two values: the request ID and the response data. Request ID
+ * is required to respond to the correct request. Use a physical connection to connect the
+ * `response` input to avoid STP violations.
```
It's a little odd to read about STP violations in the docs of a reactor, which could easily be used in unfederated programs as well. Why are any of these examples federations at all? At a quick glance, it isn't clear to me. Do they have to be?

Depetrol replied:
For the minimal +1 examples, they can be federated or unfederated and it wouldn't have much effect on the behavior of the program. I will change these examples to unfederated execution. However for the distributed logging example, we want to show the consistency-availability tradeoff that could be achieved with federated execution using `STA`.
